File: src/main/java/ca/yyx/hu/aap/protocol/MsgType.java
package ca.yyx.hu.aap.protocol;

/**
 * @author algavris
 * @date 30/11/2016.
 */

public class MsgType {

    public static final int SIZE = 2;

    public static class Control {
        public static final int MEDIADATA = 0x00;
        public static final int CODECDATA = 0x01;
        public static final int VERSIONRESPONSE = 0x02;
        public static final int SSLHANDSHAKE = 0x03;
        public static final int SERVICEDISCOVERYREQUEST = 0x05;
        public static final int SERVICEDISCOVERYRESPONSE = 0x06;
        public static final int CHANNELOPENREQUEST = 0x07;
        public static final int CHANNELOPENRESPONSE = 0x08;
        public static final int PINGREQUEST = 0x0B;
        public static final int PINGRESPONSE = 0x0C;
        public static final int NAVFOCUSREQUESTNOTIFICATION = 0x0D;
        public static final int NAVFOCUSRNOTIFICATION = 0x0E;
        public static final int BYEYEREQUEST = 0x0F;
        public static final int SHUTDOWNRESPONSE = 0x10;
        public static final int VOICESESSIONNOTIFICATION = 0x11;
        public static final int AUDIOFOCUSREQUESTNOTFICATION = 0x12;
        public static final int AUDIOFOCUSNOTFICATION = 0x13;
    }

    public static class Media {
        public static final int SETUPREQUEST = 0x8000;
        public static final int STARTREQUEST = 0x8001;
        public static final int STOPREQUEST = 0x8002;
        public static final int CONFIGRESPONSE = 0x8003;
        public static final int ACK = 0x8004;
        public static final int MICREQUEST = 0x8005;
        public static final int MICRESPONSE = 0x8006;
        public static final int VIDEOFOCUSREQUESTNOTIFICATION = 0x8007;
        public static final int VIDEOFOCUSNOTIFICATION = 0x8008;
    }

    public static class Sensor {
        public static final int STARTREQUEST = 0x8001;
        public static final int STARTRESPONSE = 0x8002;
        public static final int EVENT = 0x8003;
    }

    public static class Input {
        public static final int EVENT = 0x8001;
        public static final int BINDINGREQUEST = 0x8002;
        public static final int BINDINGRESPONSE = 0x8003;
    }


    public static String name(int type, int channel) {

        switch (type) {
            case Control.MEDIADATA:
                return "Media Data";
            case Control.CODECDATA:
                return "Codec Data";
            case Control.VERSIONRESPONSE:
                // short:major  short:minor   short:status
                return "Version Response";
            case Control.SSLHANDSHAKE:
                return "SSL Handshake Data";
            case 4:
                return "SSL Authentication Complete Notification";
            case Control.SERVICEDISCOVERYREQUEST:
                return "Service Discovery Request";
            case Control.SERVICEDISCOVERYRESPONSE:
                return "Service Discovery Response";
            case Control.CHANNELOPENREQUEST:
                return "Channel Open Request";
            case Control.CHANNELOPENRESPONSE:
                return "Channel Open Response";                       // byte:status
            case 9:
                return "9 ??";
            case 10:
                return "10 ??";
            case Control.PINGREQUEST:
                return "Ping Request";
            case Control.PINGRESPONSE:
                return "Ping Response";
            case Control.NAVFOCUSREQUESTNOTIFICATION:
                return "Navigation Focus Request";
            case Control.NAVFOCUSRNOTIFICATION:
                return "Navigation Focus Notification";               // NavFocusType
            case Control.BYEYEREQUEST:
                return "Byebye Request";
            case Control.SHUTDOWNRESPONSE:
                return "Byebye Response";
            case Control.VOICESESSIONNOTIFICATION:
                return "Voice Session Notification";
            case Control.AUDIOFOCUSREQUESTNOTFICATION:
                return "Audio Focus Request";
            case Control.AUDIOFOCUSNOTFICATION:
                return "Audio Focus Notification";                    // AudioFocusType   (AudioStreamType ?)

            case Media.SETUPREQUEST:
                return "Media Setup Request";                        // Video and Audio sinks receive this and send k3 3 / 32771
            case Media.STARTREQUEST:
                if (channel == Channel.AA_CH_SEN) {
                    return "Sensor Start Request";
                } else if (channel == Channel.AA_CH_TOU) {
                    return "Input Event";
                }
                return "Media Start Request";
            case Media.STOPREQUEST:
                if (channel == Channel.AA_CH_SEN) {
                    return "Sensor Start Response";
                } else if (channel == Channel.AA_CH_TOU) {
                    return "Input Binding Request";
                }
                return "Media Stop Request";
            case Media.CONFIGRESPONSE:
                if (channel == Channel.AA_CH_SEN) {
                    return "Sensor Event";
                } else if (channel == Channel.AA_CH_TOU) {
                    return "Input Binding Response";
                }
                return "Media Config Response";
            case Media.ACK:
                return "Codec/Media Data Ack";
            case Media.MICREQUEST:
                return "Mic Start/Stop Request";
            case Media.MICRESPONSE:
                return "Mic Response";
            case Media.VIDEOFOCUSREQUESTNOTIFICATION:
                return "Video Focus Request";
            case Media.VIDEOFOCUSNOTIFICATION:
                return "Video Focus Notification";

            case 65535:
                return "Framing Error Notification";
        }
        return "Unknown";
    }

}

File: src/main/java/ca/yyx/hu/aap/protocol/MsgType.kt
package ca.yyx.hu.aap.protocol

import ca.yyx.hu.aap.protocol.nano.Control
import ca.yyx.hu.aap.protocol.nano.Media

/**
 * @author algavris
 * *
 * @date 30/11/2016.
 */

object MsgType {

    const val SIZE = 2

    fun isControl(type: Int): Boolean {
        return type >= Control.MSG_CONTROL_MEDIADATA && type <= Control.MSG_CONTROL_AUDIOFOCUSNOTFICATION
    }

    fun name(type: Int, channel: Int): String {

        when (type) {
            Control.MSG_CONTROL_MEDIADATA -> return "Media Data"
            Control.MSG_CONTROL_CODECDATA -> return "Codec Data"
            Control.MSG_CONTROL_VERSIONRESPONSE ->
                // short:major  short:minor   short:status
                return "Version Response"
            Control.MSG_CONTROL_HANDSHAKE -> return "SSL Handshake Data"
            4 -> return "SSL Authentication Complete Notification"
            Control.MSG_CONTROL_SERVICEDISCOVERYREQUEST -> return "Service Discovery Request"
            Control.MSG_CONTROL_SERVICEDISCOVERYRESPONSE -> return "Service Discovery Response"
            Control.MSG_CONTROL_CHANNELOPENREQUEST -> return "Channel Open Request"
            Control.MSG_CONTROL_CHANNELOPENRESPONSE -> return "Channel Open Response"                       // byte:status
            9 -> return "9 ??"
            10 -> return "10 ??"
            Control.MSG_CONTROL_PINGREQUEST -> return "Ping Request"
            Control.MSG_CONTROL_PINGRESPONSE -> return "Ping Response"
            Control.MSG_CONTROL_NAVFOCUSREQUESTNOTIFICATION -> return "Navigation Focus Request"
            Control.MSG_CONTROL_NAVFOCUSRNOTIFICATION -> return "Navigation Focus Notification"               // NavFocusType
            Control.MSG_CONTROL_BYEYEREQUEST -> return "Byebye Request"
            Control.MSG_CONTROL_BYEYERESPONSE -> return "Byebye Response"
            Control.MSG_CONTROL_VOICESESSIONNOTIFICATION -> return "Voice Session Notification"
            Control.MSG_CONTROL_AUDIOFOCUSREQUESTNOTFICATION -> return "Audio Focus Request"
            Control.MSG_CONTROL_AUDIOFOCUSNOTFICATION -> return "Audio Focus Notification"                    // AudioFocusType   (AudioStreamType ?)

            Media.MSG_MEDIA_SETUPREQUEST -> return "Media Setup Request"                        // Video and Audio sinks receive this and send k3 3 / 32771
            Media.MSG_MEDIA_STARTREQUEST -> {
                if (channel == Channel.ID_SEN) {
                    return "Sensor Start Request"
                } else if (channel == Channel.ID_INP) {
                    return "Input Event"
                } else if (channel == Channel.ID_MPB) {
                    return "Media Playback Status"
                }
                return "Media Start Request"
            }
            Media.MSG_MEDIA_STOPREQUEST -> {
                if (channel == Channel.ID_SEN) {
                    return "Sensor Start Response"
                } else if (channel == Channel.ID_INP) {
                    return "Input Binding Request"
                } else if (channel == Channel.ID_MPB) {
                    return "Media Playback Status"
                }
                return "Media Stop Request"
            }
            Media.MSG_MEDIA_CONFIGRESPONSE -> {
                if (channel == Channel.ID_SEN) {
                    return "Sensor Event"
                } else if (channel == Channel.ID_INP) {
                    return "Input Binding Response"
                } else if (channel == Channel.ID_MPB) {
                    return "Media Playback Status"
                }
                return "Media Config Response"
            }
            Media.MSG_MEDIA_ACK -> return "Codec/Media Data Ack"
            Media.MSG_MEDIA_MICREQUEST -> return "Mic Start/Stop Request"
            Media.MSG_MEDIA_MICRESPONSE -> return "Mic Response"
            Media.MSG_MEDIA_VIDEOFOCUSREQUESTNOTIFICATION -> return "Video Focus Request"
            Media.MSG_MEDIA_VIDEOFOCUSNOTIFICATION -> return "Video Focus Notification"

            65535 -> return "Framing Error Notification"
        }
        return "Unknown"
    }

}
